src/index.ts

```diff
@@ -142,54 +142,64 @@ export async function* getDeployedEntitiesStream(
-        components,
-        hash,
-        options.tmpDownloadFolder,
-        [options.contentServer],
-        new Map(),
-        options.requestMaxRetries,
-        options.requestRetryWaitTime
-      )
+  for (const snapshot of snapshots) {
+    const { hash, lastIncludedDeploymentTimestamp, replacedSnapshotHashes } = snapshot
+    logs.debug(`Processing snapshot: ${snapshot.hash} from ${options.contentServer}`)
+    // 2. for each snapshot, download the snapshot file if it contains deployments
+    //    in the range we are interested (>= genesisTimestamp)
+    if (
+      hash && lastIncludedDeploymentTimestamp && lastIncludedDeploymentTimestamp > genesisTimestamp &&
+      !(await components.processedSnapshotStorage.wasSnapshotProcessed(hash, replacedSnapshotHashes))) {
+      try {
+        // 2.1. download the snapshot file if needed
+        await downloadFileWithRetries(
+          components,
+          hash,
+          options.tmpDownloadFolder,
+          [options.contentServer],
+          new Map(),
+          options.requestMaxRetries,
+          options.requestRetryWaitTime
+        )
 
-      // 2.2. open the snapshot file and process line by line
-      const deploymentsInFile = processDeploymentsInFile(hash, components)
-      for await (const rawDeployment of deploymentsInFile) {
-        const deployment = coerceEntityDeployment(rawDeployment)
-        if (!deployment) continue
-        // selectively ignore deployments by localTimestamp
-        if (deployment.localTimestamp >= genesisTimestamp) {
-          components.metrics.increment('dcl_entities_deployments_processed_total', metricLabels)
-          yield deployment
-        }
-        // update greatest processed timestamp
-        if (deployment.localTimestamp > greatestProcessedTimestamp) {
-          greatestProcessedTimestamp = deployment.localTimestamp
+        // 2.2. open the snapshot file and process line by line
+        const deploymentsInFile = processDeploymentsInFile(hash, components)
+        for await (const rawDeployment of deploymentsInFile) {
+          const deployment = coerceEntityDeployment(rawDeployment)
+          if (!deployment) continue
+          // selectively ignore deployments by localTimestamp
+          if (deployment.localTimestamp >= genesisTimestamp) {
+            components.metrics.increment('dcl_entities_deployments_processed_total', metricLabels)
+            yield deployment
+          }
+          // update greatest processed timestamp
+          if (deployment.localTimestamp > greatestProcessedTimestamp) {
+            greatestProcessedTimestamp = deployment.localTimestamp
+          }
         }
-      }
-    } finally {
-      if (options.deleteSnapshotAfterUsage !== false) {
-        try {
-          await components.storage.delete([hash])
-        } catch (err: any) {
-          logs.error(err)
+        await components.processedSnapshotStorage.markSnapshotProcessed(hash, replacedSnapshotHashes)
```

Comment:
This line does not ensures that all the entities were "deployed" right? only sent to the next stage of the pipeline (p-queue), which will later deploy them.

Does this logic risk losing deployments? Assume that the queue fails or we restart the service while it is draining. Will those deployments be lost?

1. Read snapshot bafky1
2. Send all 1000 entities to the queue (yield)
3. Mark bafky1 as read
4. Deploy up to 101 entities from the queue
5. *SERVICE RESTART*
6. ~Read snapshot bafky1~ -> IGNORE, bafky1 was already processed
7. *899 deployments lost in the limbo*

Reply:
This is correct. This logic risks losing deployments. 😭

Reply:
Aren't we storing the failed deployments to avoid that?

Reply:
the failed deployments are stored upon failure, not when added to the queue

Reply:
Yes, but the restart could happen before a deployment is done (failed or not)

Reply:
This issue was resolved. When the snapshots-fetcher streams entities, it sends a callback that the Catalyst uses when the entity was effectively deployed or persisted as failed deployed. The batch deployer gets an element from the queue, effectively deploys or persists as failed deployment and at the end calls the callback sent by the snapshot-fetcher saying "effectively deployed an entity in this snapshot".